### PR TITLE
Batch methods for MapQuest & Bing

### PR DESCRIPTION
**Pull request by ThmsLa**
Pretty much self-explaining.

- Mapquest Batch to send and retrieve 100 geocodings at once
- Bing Batch for 2500. Though... The Bing batch method is quite dirty as it needs to follow the [Bing workflow](https://msdn.microsoft.com/en-us/library/ff701733.aspx). 

Would love your review on these, especially if you have an idea of a cleaner way for the Bing one 😄

### CODE REVIEW

**Comment by DenisCarriere** [comment]
@ThmsLa Great work, unfortunately I wouldn't have some great advice to give you for the BingBatch... I've also attempted to create a batch request with little success. Therefor I'm sure your attempt is better then mine.

Sadly some of the tests are failing, however it's not because of you.

image


I'm willing to merge this 👍 

@ebreton any thoughts?

**Comment by ebreton** [comment]
Hi Guys! @ThmsLa thanks for the PR! WIll have a look later today. 😁

File: geocoder/mapquest_reverse.py [hunk 1]
```diff
@@ -31,9 +31,11 @@ class MapquestReverse(MapquestQuery):
     provider = 'mapquest'
     method = 'reverse'
 
+    _URL = 'http://www.mapquestapi.com/geocoding/v1/reverse'
+
     def _build_params(self, location, provider_key, **kwargs):
         return {
-            'key': self._get_api_key(mapquest_key),
+            'key': provider_key,
```
Comment (ebreton) [inline geocoder/mapquest_reverse.py]: interesting that you need to shortcut the default behaviour here... what was the motivation ?

Comment (ThmsLa) [reply geocoder/mapquest_reverse.py]: Correct me if I'm wrong, but I believe that was a bug:

Calling it here (in mapquest_reverse.py) was preventing users from calling the method with the key as an argument.
Furthermore the base class (MultipleResultsQuery) already calls \_get_api_key() in its \_\_init__ method, ensuring that either a key was passed as an argument or that it is available through environment variables.

**Comment by ThmsLa** [comment]
After going through my code again, I thought that some explanations might help you guys as the Bing workflow is quite strange.

There are 4 steps to batch geocode using it, which are:
- preparing the data for upload in their CSV format, which is what the `generate_batch()` function is doing.
- upload that file to their API, which will in return give you an ID to check if that job has finished. This ID is extracted in `extract_resource_id()`.
- check if their service has finished processing the data. The `is_job_done()` function takes care of that, and is called every 5 seconds (`_BATCH_WAIT`) for one minute (`_BATCH_TIMEOUT`) which is when we consider that the job has failed.
- retrieve the geocoded data through `get_job_result()`.

If you need any other explanations, please let me know @ebreton and @DenisCarriere :)

File: geocoder/mapquest_reverse.py (hunk 1 above)
Comment (ebreton) [reply geocoder/mapquest_reverse.py]: yes you are perfectly right 👍

**Comment by ebreton** [comment]
As @DenisCarriere pointed it out, the tests are failing for other reasons than this PR. I am thus merging it and will look at fixing the tests in another request.

@ThmsLa : thanks a lot for the PR ! and the additionnal pieces of information 🎉 

🎈

**Pull request merged**